app/Main.hs

```diff
@@ -66,82 +68,25 @@ configFromEnv = do
   where
     noEnv env = env ++ " environment variable does not exist"
 
-data HapCommand
-  = HapDeploy
-  | HapRollback
-  deriving Show
-
-parseHapCommand :: String -> Maybe HapCommand
-parseHapCommand "deploy" = Just HapDeploy
-parseHapCommand "rollback" = Just HapRollback
-parseHapCommand _ = Nothing
-
-data HapOptions =
-  HapOptions
-    { hapCommand :: Maybe HapCommand
-    , hapHelp :: Bool
-    }
-  deriving Show
-
-defaultHapOptions :: HapOptions
-defaultHapOptions =
-  HapOptions
-    { hapCommand = Nothing
-    , hapHelp = False
-    }
-
-hapOptionDescriptions :: [GetOpt.OptDescr (HapOptions -> HapOptions)]
-hapOptionDescriptions =
-  [ GetOpt.Option
-      ['h']
-      ["help"]
-      (GetOpt.NoArg (\hapOptions -> hapOptions { hapHelp = True }))
-      "Show this help text"
-
-  ]
-
-parseHapOptions :: [String] -> Either String HapOptions
-parseHapOptions args =
-  case GetOpt.getOpt GetOpt.Permute hapOptionDescriptions args of
-    (options, [], []) ->
-      Right (foldl (flip id) defaultHapOptions options)
-
-    (options, [command], []) ->
-      case parseHapCommand command of
-        Nothing ->
-          Left ("Invalid argument: " ++ command)
-
-        maybeHC ->
-          Right (foldl (flip id) defaultHapOptions {hapCommand = maybeHC} options)
-
-    _ ->
-      Left "First argument must be either 'deploy' or 'rollback'."
-
-hapHelpAction :: Maybe HapCommand -> IO ()
-hapHelpAction _ =
-  putStrLn hapUsage >> Exit.exitSuccess
-
-hapUsage :: String
-hapUsage =
-  GetOpt.usageInfo hapUsageHeader hapOptionDescriptions
-
-hapUsageHeader :: String
-hapUsageHeader =
-  "usage: hap [-h | --help] <command>\n"
-
 main :: IO ()
 main = do
-  eitherHapOptions <- fmap parseHapOptions Environment.getArgs
-
-  HapOptions{..} <- either Exit.die return eitherHapOptions
-
-  Monad.when hapHelp (hapHelpAction hapCommand)
-
   hapConfiguration <- configFromEnv
+  Monad.join $ Opt.customExecParser prefs (Opt.info (opts hapConfiguration) Opt.idm)
+    where
+      opts hapConfig =
+        Opt.helper
+          Opt.<*> Opt.subparser (commands hapConfig)
+          Opt.<|> Opt.flag' printVersion (Opt.long "version" <> Opt.short 'v' <> Opt.help "Diplay the version of Hapistrano")
+      commands hapConfig =
+        Opt.command "deploy"
+          (Opt.info (Opt.pure $ deploy hapConfig) (Opt.progDesc "Deploys the current release with the configure options"))
+        <> Opt.command "rollback"
+          (Opt.info (Opt.pure $ rollback hapConfig) (Opt.progDesc "Rolls back to the previous release"))
+
+prefs :: ParserPrefs
+prefs =
+  Opt.defaultPrefs { prefShowHelpOnEmpty = True }
```

Comment:
I agree with you. I don't like to set the env variables to then see the help information. I'll see what can I do about it. Thanks @jpvillaisaza